libraries/chain/include/eos/chain/authority_checker.hpp

```diff
@@ -0,0 +1,127 @@
+#pragma once
+
+#include <eos/chain/types.hpp>
+#include <eos/types/generated.hpp>
+
+#include <eos/utilities/parallel_markers.hpp>
+
+#include <fc/scoped_exit.hpp>
+
+#include <boost/range/algorithm/find.hpp>
+#include <boost/algorithm/cxx11/all_of.hpp>
+
+namespace eos { namespace chain {
+
+namespace detail {
+using MetaPermission = static_variant<types::KeyPermissionWeight, types::AccountPermissionWeight>;
+
+struct GetWeightVisitor {
+   using result_type = UInt32;
+
+   template<typename Permission>
+   UInt32 operator()(const Permission& permission) { return permission.weight; }
+};
+
+// Orders permissions descending by weight, and breaks ties with Key permissions being less than Account permissions
+struct MetaPermissionComparator {
+   bool operator()(const MetaPermission& a, const MetaPermission& b) const {
+      GetWeightVisitor scale;
+      if (a.visit(scale) > b.visit(scale)) return true;
+      return a.contains<types::KeyPermissionWeight>() && !b.contains<types::KeyPermissionWeight>();
+   }
+};
+
+using MetaPermissionSet = boost::container::flat_multiset<MetaPermission, MetaPermissionComparator>;
+}
+
+/**
+ * @brief This class determines whether a set of signing keys are sufficient to satisfy an authority or not
+ *
+ * To determine whether an authority is satisfied or not, we first determine which keys have approved of a message, and
+ * then determine whether that list of keys is sufficient to satisfy the authority. This class takes a list of keys and
+ * provides the @ref satisfied method to determine whether that list of keys satisfies a provided authority.
+ *
+ * @tparam F A callable which takes a single argument of type @ref AccountPermission and returns the corresponding
+ * authority
+ */
+template<typename F>
+class AuthorityChecker {
+   F PermissionToAuthority;
+   vector<public_key_type> signingKeys;
+   vector<bool> usedKeys;
+
+   struct WeightTallyVisitor {
```

Comment:
> I would consider making the visitor simply produce the weight and then putting the aggregation of weights in the ::satisfied method where the result is important/used.

Not sure I understand what you mean here. Do you mean not taking the early out you mentioned below?